ec2/utility.py

```diff
@@ -0,0 +1,120 @@
+
+
+def validate_state(instance, state, timeout_length, check_interval):
+    """ Check if an EC2 instance is in a particular state.
+
+    :param instance: And EC2 instance.
+    :param state: The state code (pending = 0, running = 16,
+                  shutting down = 32, terminated = 48, stopping = 64,
+                  stopped = 80
+    :param timeout_length: How long to wait for a positive answer
+           before we stop checking.
+    :param check_interval: How long to wait between checks.
+    :return: bool (True the desired state was reached, False, it was not.)
+    """
+
+    ctx.logger.debug('(Node: {0}): Attempting state validation: '
+                     'instance id: {0}, state: {1}, timeout length: {2}, '
+                     'check interval: {3}.'.format(instance.id, state,
+                                                   timeout_length,
+                                                   check_interval))
+
+    if check_interval < 1:
+        check_interval = 1
+
+    timeout = time.time() + timeout_length
+
+    while True:
+        if state == get_instance_state(instance):
+            ctx.logger.info('(Node: {0}): '
+                            'Instance state validated: instance {0}.'
+                            .format(instance.state))
+            return True
+        elif time.time() > timeout:
+            raise NonRecoverableError('(Node: {0}): Timed out during'
+                                      'instance state validation: '
+                                      'instance: {1}, '
+                                      'timeout length: {2}, '
+                                      'check interval: {3}.'
+                                      .format(ctx.instance.id,
+                                              timeout_length,
+                                              check_interval))
+        time.sleep(check_interval)
+
+
+def get_instance_state(instance, ctx):
+    """
+
+    :param instance:
+    :return:
+    """
+
+    state = instance.update()
+    ctx.logger.debug('(Node: {0}): Instance state is {1}.'
+                     .format(ctx.instance.id, state))
+    return instance.state_code
+
+
+def handle_ec2_error(ctx_instance_id, ec2_error, action, ctx):
+    """
+
+    :param ctx_instance_id: the Cloudify Context node ID
+    :param ec2_error: EC2 Error Object
+    :param action: A string that fits in nicely with the error code :)
+    """
+
+    raise NonRecoverableError('(Node: {0}): Error. '
+                              'Failed to {1} instance: '
+                              'API returned: {2}.'
+                              .format(ctx_instance_id, action, ec2_error))
+
+
+def validate_instance_id(instance_id, ctx):
+    """
+
+    :param instance_id: An EC2 instance ID
+    :return: True that the instance ID is valid or
+             throws unrecoverable error
+    """
+
+    try:
+        reservations = EC2().get_all_reservations(instance_id)
+    except EC2ResponseError:
```

Comment:
this can be replaced with something like this:

```python
except (EC2ResponseError, BotoServerError) as e:
    raise NonRecoverableError('...'.format(e))
```

so maybe you can drop the handle_ec2_error() function.